src/wsdl.rs

```diff
@@ -227,83 +229,133 @@ impl<'a, 'input> WsPortOperation<'a, 'input> {
-            None => return Ok(None),
-        };
-
-        let (_message_namespace, message_name) =
-            split_qualified(message_typename).map_err(|e| WsError::new(self.0, e))?;
+            .filter(|n| n.has_tag_name(("http://schemas.xmlsoap.org/wsdl/", "output")))
+            .filter_map(|n| Some((n.attribute("name")?, n.attribute("message")?)))
+            .filter_map(
+                |(name, message_typename)| match split_qualified(message_typename) {
+                    Ok((_message_namespace, message_name)) => Some((name, message_name)),
+                    Err(_) => None,
+                },
+            )
+            .filter_map(move |(name, message_name)| {
+                let Ok(mut messages) = def.messages() else {
+                    return None;
+                };
+                if let Some(message) =
+                    messages.find(|n| n.0.attribute("name") == Some(message_name))
+                {
+                    Some((name, message))
+                } else {
+                    None
+                }
+            }))
+    }
+
+    /// Retrieve the first fault message for this port.
+    pub fn fault(&self) -> Result<Option<WsMessage<'a, 'input>>> {
+        let mut faults = self.faults()?;
+        let fault = faults.next();
+        if faults.next().is_some() {
+            panic!(
+                "Multiple fault messages found for operation {:?}",
+                self.name()?
+            );
+        }
+        Ok(fault.map(|(_, m)| m))
+    }
 
+    /// Retrieve all fault messages for this port
+    pub fn faults(&self) -> Result<impl Iterator<Item = (&'a str, WsMessage<'a, 'input>)>> {
         let def = WsDefinitions::find_parent(self.0)?;
-        Ok(Some(
-            def.messages()?
-                .find(|n| n.0.attribute("name") == Some(message_name))
-                .ok_or(WsError::new(
-                    self.0,
-                    WsErrorType::InvalidReference(message_name.to_string()),
-                ))?,
-        ))
+        Ok(self
+            .0
+            .children()
+            .filter(|n| n.has_tag_name(("http://schemas.xmlsoap.org/wsdl/", "fault")))
+            .filter_map(|n| Some((n.attribute("name")?, n.attribute("message")?)))
+            .filter_map(
+                |(name, message_typename)| match split_qualified(message_typename) {
+                    Ok((_message_namespace, message_name)) => Some((name, message_name)),
+                    Err(_) => None,
+                },
+            )
+            .filter_map(move |(name, message_name)| {
+                let Ok(mut messages) = def.messages() else {
+                    return None;
+                };
+                if let Some(message) =
+                    messages.find(|n| n.0.attribute("name") == Some(message_name))
+                {
+                    Some((name, message))
+                } else {
+                    None
+                }
+            }))
     }
```

Comment:
I'd suggest avoiding excessive iterator chaining and instead do something like this:
```rs
    /// Retrieve the fault message for this port.
    pub fn faults(&self) -> Result<impl Iterator<Item = WsMessage<'a, 'input>>> {
        let message_typenames = self
            .0
            .children()
            .filter(|n| n.has_tag_name(("http://schemas.xmlsoap.org/wsdl/", "fault")))
            .map(|n| n.attribute("message"))
            .flatten();

        let mut messages = Vec::new();

        for message_typename in message_typenames {
            let (_message_namespace, message_name) =
                split_qualified(message_typename).map_err(|e| WsError::new(self.0, e))?;

            let def = WsDefinitions::find_parent(self.0)?;
            if let Some(message) = def
                .messages()?
                .find(|n| n.0.attribute("name") == Some(message_name))
            {
                messages.push(message);
            }
        }

        Ok(messages.into_iter())
    }
```

That way you can surface any errors encountered, rather than silently eliminating elements :)